h2o-algos/src/main/java/hex/tree/TreeHandler.java

```diff
@@ -504,11 +528,23 @@ private static void fillNodeSplitTowardsChildren(final StringBuilder nodeDescrip
         return null;
     }
 
-    private static void fillNodeCategoricalSplitDescription(final StringBuilder nodeDescriptionBuilder, final SharedTreeNode node) {
+    private static String[] namesForNodelevels(final int[] indices, final String[] domainValues) {
+        final String[] names = new String[indices.length];
+        for (int i = 0; i < indices.length; i++) {
+            names[i] = domainValues[indices[i]];
+        }
+
+        return names;
+    }
+
+    private static void fillNodeCategoricalSplitDescription(final StringBuilder nodeDescriptionBuilder, final SharedTreeNode node,
+                                                            String[][] leftCatSplits, final String[][] rightCatSplits, final int pointer) {
         final SharedTreeNode leftChild = node.getLeftChild();
         final SharedTreeNode rightChild = node.getRightChild();
         final int[] leftChildLevels = extractNodeLevels(leftChild);
         final int[] rightChildLevels = extractNodeLevels(rightChild);
+        leftCatSplits[pointer] = namesForNodelevels(leftChildLevels, node.getDomainValues());
```

Comment:
Not much complexity added, this is done in-place of the original description builder. Only some more args are passed in to allow these to be calculated as well.